File: 100-doors/C++/solution.cpp
#include <iostream>

int main()
{
  int square = 1, increment = 3;
  for (int door = 1; door <= 100; ++door)
  {
    std::cout << "Door " << door;
    if (door == square)
    {
      std::cout << " is Open" << std::endl;
      square += increment;
      increment += 2;
    }
    else
      std::cout << " is Closed" << std::endl;
  }
  return 0;
}

File: 100-doors/Haskell/solution.hs
data Door
  = Open
  | Closed
  deriving (Eq, Show)

toggle :: Door -> Door
toggle Open = Closed
toggle Closed = Open

toggleEvery :: Int -> [Door] -> [Door]
toggleEvery k = zipWith toggleK [1 ..]
  where
    toggleK n door
      | n `mod` k == 0 = toggle door
      | otherwise = door

run :: Int -> [Door]
run n = foldr toggleEvery (replicate n Closed) [1 .. n]

main :: IO ()
--main = print $ (map (\(x,y) -> putStrLn $ "Door " ++  show (x) ++ " is " ++ show (y) ++ "\n" ) (zip [1 ..] (run 100)))
main = print $ (map (\(x,y) -> "Door " ++  show (x) ++ " is " ++ show (y) ++ "\n" ) (zip [1 ..] (run 100)))
